File: 프로그래머스/2/84512. 모음 사전/모음 사전.swift
// 리턴 - 몇번째 단어?
// 모든 경우의 수를 넣고, 정렬 -> 3000개 정도

import Foundation

func solution(_ word:String) -> Int {
    
    let n = word.count
    let chars: [String] = ["A", "E", "I", "O", "U"]
    var dict = Set<String>()
    
    func dfs(_ char: String) {
        if char.count == 5 { return }
        
        for c in chars {
            let newChar = char + c
            dict.insert(newChar)
            dfs(newChar)
        }
    }
    dfs("")
    
    let arr = Array(dict).sorted()
    
    for i in 0..<arr.count {
        if arr[i] == word { return i + 1 }
    }
    
    return 0
}

File: 프로그래머스/2/84512. 모음 사전/모음 사전.py
# 모음으로 만드는 길이 5 이하의 모든 단어
# 해당 단어는 몇번째 단어?

# 1~5개 단어
# 5P1 + 5P2 + 5P3 + 5P4 + 5P5 = 5 + 25 + 125 + 625 + 3125 > 대강 4000개

# 단어부터 만든다음 > 일치여부 확인해도 될 것 같은데

def solution(word):
    
    collection = []
    
    for i in range(1, 6):
        for w in permu(i):
            collection.append(w)
            
    collection.sort()
    
    for i in range(len(collection)):
        if collection[i] == word:
            return i+1


chars = ['A', 'E', 'I', 'O', 'U']

# 중복 순열
def permu(n):

    result = set()
    cur = ""
    
    def select(k):
        nonlocal cur
        
        if k == n:
            result.add(cur)
            return
        
        for w in chars:
            cur += w
            select(k+1)
            cur = cur[:-1]
            
    select(0)
    
    return result
    